pkg/ccl/streamingccl/streamclient/partitioned_stream_client.go

```diff
@@ -0,0 +1,221 @@
+	if row.Err() != nil {
+		return nil, errors.Wrap(row.Err(), "Error in planning a replication stream")
+	}
+
+	var rawSpec []byte
+	if err = row.Scan(&rawSpec); err != nil {
+		return nil, err
+	}
+	var spec streampb.ReplicationStreamSpec
+	if err := protoutil.Unmarshal(rawSpec, &spec); err != nil {
+		return nil, err
+	}
+
+	// we need to use the disqlplanner to find all the planner nodes and
+	// assign partition to it in a round-robin fashion
+	topology := Topology{}
+	for _, p := range spec.Partitions {
+		rawSpec, err := protoutil.Marshal(p.PartitionSpec)
+		if err != nil {
+			return nil, err
+		}
+		topology = append(topology, PartitionInfo{
+			ID:                p.NodeID.String(), // how do we determine partition id?
+			SubscriptionToken: SubscriptionToken(rawSpec),
+			SrcInstanceID:     int(p.NodeID),
+			SrcAddr:           streamingccl.PartitionAddress(p.SQLAddress.String()),
+			SrcLocality:       p.Locality,
+		})
+	}
+	return topology, nil
+}
+
+type partitionStreamDecoder struct {
+	rows *gosql.Rows
+	e    streampb.StreamEvent
+}
+
+func (d *partitionStreamDecoder) pop() streamingccl.Event {
+	if d.e.Checkpoint != nil {
+		event := streamingccl.MakeCheckpointEvent(d.e.Checkpoint.Spans[0].Timestamp)
+		d.e.Checkpoint = nil
+		return event
+	}
+
+	if d.e.Batch != nil {
+		event := streamingccl.MakeKVEvent(d.e.Batch.KeyValues[0])
+		d.e.Batch.KeyValues = d.e.Batch.KeyValues[1:]
+		if len(d.e.Batch.KeyValues) == 0 {
+			d.e.Batch = nil
+		}
+		return event
+	}
+
+	return nil
+}
+
+func (d *partitionStreamDecoder) decode() error {
+	var data []byte
+	if err := d.rows.Scan(&data); err != nil {
+		return err
+	}
+	var streamEvent streampb.StreamEvent
+	if streamEvent.Checkpoint == nil && streamEvent.Batch == nil {
+		return errors.New("unexpected event type")
+	}
+	d.e = streamEvent
+	return nil
+}
+
+func (p *partitionedStreamClient) Subscribe(
+	ctx context.Context, stream StreamID, spec SubscriptionToken, checkpoint hlc.Timestamp,
+) (chan streamingccl.Event, chan error, error) {
+
+	db, err := gosql.Open("postgres", p.remote.String())
+	if err != nil {
+		return nil, nil, err
+	}
+	conn, err := db.Conn(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	sps := streampb.StreamPartitionSpec{}
+	if err = protoutil.Unmarshal(spec, &sps); err != nil {
+		return nil, nil, err
+	}
+	sps.StartFrom = checkpoint
+
+	specBytes, err := protoutil.Marshal(&sps)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	rows, err := conn.QueryContext(ctx, `SELECT * FROM crdb_internal.stream_partition($1)`, specBytes)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	eventChan := make(chan streamingccl.Event)
+	errChan := make(chan error)
+	// Spin up a goroutine to read rows in partition stream
+	go func() {
+		psd := partitionStreamDecoder{rows: rows}
+		for rows.Next() {
+			if err := psd.decode(); err != nil {
+				errChan <- err
+				break
+			}
+			for e := psd.pop(); e != nil; {
+				eventChan <- e
+			}
```

Comment:
This is almost certainly not what we should do.
First, spinning up go routine like this is probably not okay.  You want the caller to be able to control the lifetime of that go routine -- by using something like context group  (so that it can be cancelled, shutdown, and waited for).

Secondly, we definitely do not want to have "pop" like interface like I had in the test -- the reason is you want to send those batches directly to the caller -- perhaps those batches contain whole SStables, etc,.  Let the caller manage this and figure out what to do.  I understand that right now, the rest of the code assumes that events represent a single event -- but that really has to change.

I think this method should be changed to return a `Subscription` (either a struct or interface impl).
Something like this:

```
type Subscription interface {
  // Starts receiving subscription events.  Terminates when context
  // cancelled.
  Receive(ctx context.Context) error

  // Events returns a channel receiving stream events.
  Events() <-chan streamingccl.Event

  // Err returns a channel receiving error events
 Err() <-chan error
}
```

Receive -- connects to the database, creates your rows, etc and just starts running a *blocking* loop:
```
  for rows.Next() {
    // decode; push to events or err channel
  }
```

I think this would allow the caller to do something like this:

```
subscription := client.Subsribe()
ctxGroup := ctxtGroup(ctx)
ctxGroup.GoCtx(subscription.Receive)
ctxGroup.GoCtx(func(ctx context.Context, sub Subscription) error {
    return processEvents(ctx, sub)
}
....
func processEvents {
   checkpointTimer := newTimer...
   for {
      select {
        case ctx.Done()...
        case e := sub.Events(): ...
        case err := sub.Err(): ..
        case checkpontTimer:
            heartbeat()....
        etc...
   }
}
```